File: 1046-last-stone-weight/max-heap/java/solution.java
class Solution {
    public int lastStoneWeight(int[] stones) {
        // Java's PriorityQueue is min-heap by default. Passing Collections.reverseOrder() turns it into max-heap.
        PriorityQueue<Integer> maxHeap = new PriorityQueue<>(Collections.reverseOrder());

        for (int stone : stones) {
            maxHeap.offer(stone);
        }

        while (maxHeap.size() > 1) {
            int firstStone = maxHeap.poll();
            int secondStone = maxHeap.poll();
            if (firstStone != secondStone) {
                maxHeap.add(firstStone - secondStone);
            }
        }
        return maxHeap.isEmpty() ? 0 : maxHeap.peek();
    }
}

File: 1046-last-stone-weight/max-heap/cpp/solution.cpp
class Solution {
public:
    int lastStoneWeight(vector<int>& stones) {
        // priority_queue in C++ is max-heap by default
        priority_queue<int> max_heap(stones.begin(), stones.end());
        
        while (max_heap.size() > 1) {
            int first_stone = max_heap.top();
            max_heap.pop();
            
            int second_stone = max_heap.top();
            max_heap.pop();
            
            if (first_stone != second_stone) {
                max_heap.push(first_stone - second_stone);
            }
        }
        return max_heap.empty() ? 0 : max_heap.top();
    }
};
